### DISCUSSION TITLE
Download + Stream multimedia at the same time

### DISCUSSION THREAD

**Discussion by Dukler** [anacrolix/torrent#821] [Q&A] [answered]
I was wondering if there was a way to use this client to download music/video and at the same time being able to reproduce it/serve it.
At the time I'm only able to wait for the files to finish since until then they are locked r/w.
Thanks for the help!

**Comment by anacrolix** [comment]
Yes, that's the default behaviour, and primary use case for the implementation. Obtain a Reader to the file you want, and extract it however you wish and the client will prioritize data that you are blocked on (and do readahead). Typically you serve this through a HTTP endpoint, and your consuming application (such as a media player) makes requests that drive it.

**Reply by Dukler** [reply]
Yes, I just went and tried with a different torrent because, I have recently figured out that there are problematic extensions to do this, like .mp4 that has its "'moov' atom"(metadata) at the end of the file, and so it's very hard to reproduce/encode it, unless you don't have the entirety of it while being processed. Tried it with an ".mkv" with no trouble.

**Reply by anacrolix** [reply, marked as answer]
Yeah it's a shame, and not easy to work around some file formats. See https://en.wikipedia.org/wiki/Progressive_download. Generally I find MP4, webp and mkv are best, sometimes you have to fiddle the headers a bit. A smart player won't care, it will seek to the end, retrieve the necessary frames, then resume at the start.
